File: 2.add-two-numbers/main.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
function ListNode(val) {
    this.val = val;
    this.next = null;
}



function genList(l) {
    let header = new ListNode(0)
    let curr = header
    for (var i = 0; i < l.length; i++) {
        curr.next = new ListNode(l[i])
        curr = curr.next
    }
    return header.next
}

function showList(l) {
    do {
        console.log(l.val)
        l = l.next
    } while (l != null)
}


/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    let header = new ListNode(0)
    let curr = header
    let val = 0
    do {
        let sum = 0
        if (l1 != null) {
            sum += l1.val
            l1 = l1.next
        }
        if (l2 != null) {
            sum += l2.val
            l2 = l2.next
        }
        sum += val
        curr.next = new ListNode(sum % 10)
        curr = curr.next
        val = parseInt(sum / 10)
    } while (val != 0 || l1 != null || l2 != null)
    return header.next
};

(function main() {
    a = genList([2, 4, 3])
    b = genList([5, 6, 4])
    c = addTwoNumbers(a, b)
    showList(c)
}());

File: 2.add-two-numbers/main.go
package main

import "fmt"

/**
 * Definition for singly-linked list.

 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	header := &ListNode{}
	curr := header
	val := 0
	for true {
		if val == 0 && l1 == nil && l2 == nil {
			break
		}
		sum := 0
		if l1 != nil {
			sum = l1.Val + sum
			l1 = l1.Next
		}

		if l2 != nil {
			sum = l2.Val + sum
			l2 = l2.Next
		}
		sum = sum + val
		curr.Next = &ListNode{
			Val: sum % 10,
		}
		curr = curr.Next

		val = sum / 10

	}

	return header.Next
}

func genList(l []int) *ListNode {
	header := &ListNode{
		Val: 0,
	}
	curr := header

	for _, v := range l {
		curr.Next = &ListNode{
			Val: v,
		}
		curr = curr.Next
	}
	return header.Next

}

func showList(l *ListNode) {
	for l != nil {
		fmt.Println(l.Val)
		l = l.Next
	}
}
func main() {
	a := genList([]int{2, 4, 3})
	b := genList([]int{5, 6, 4})
	c := addTwoNumbers(a, b)
	showList(c)
}
